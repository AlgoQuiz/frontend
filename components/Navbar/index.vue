<template>
  <BNavbar toggleable="lg" type="dark" variant="primary">
    <BNavbarBrand to="/">AlgoRace</BNavbarBrand>

    <BNavbarToggle target="nav-collapse"></BNavbarToggle>

    <BCollapse id="nav-collapse" is-nav>
      <BNavbarNav v-if="isLoggedIn" class="ml-auto">
        <BNavItem to="/tasks">Tasks</BNavItem>
        <BNavItem @click="signOut">Sign out</BNavItem>
      </BNavbarNav>
      <BNavbarNav v-else class="ml-auto">
        <BNavItem to="/sign-in">Sign in</BNavItem>
        <BNavItem to="/sign-up">Sign up</BNavItem>
      </BNavbarNav>
    </BCollapse>
  </BNavbar>
</template>

<script>
import { createNamespacedHelpers } from 'vuex'

const {
  mapActions: mapUserActions,
  mapGetters: mapUserGetters,
} = createNamespacedHelpers('user')

export default {
  name: 'Navbar',

  computed: {
    ...mapUserGetters(['isLoggedIn']),
  },

  methods: {
    ...mapUserActions(['signOut']),
  },
}
</script>

<style scoped>
.nav-link.button {
  border: none;
  background: transparent !important;
}
</style>
