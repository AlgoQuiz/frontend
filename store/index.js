// import type { GetterTree, ActionTree, MutationTree } from 'vuex'
// export interface AppState {}

// export interface RootState {}

// export const state = (): RootState => ({})

// export const getters: GetterTree<RootState, RootState> = {}

// export const mutations: MutationTree<RootState> = {}

// export const actions: ActionTree<RootState, RootState> = {}

import Vue from 'vue'
import Vuex from 'vuex'

import * as movies from './movies'

Vue.use(Vuex)

export const store = new Vuex.Store({
  state: {},
  mutations: {},
  actions: {},
  modules: {
    movies: {
      namespaced: true,
      ...movies,
    },
  },
})
