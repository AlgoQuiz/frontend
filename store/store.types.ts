import { Store } from 'vuex/types/index'

import { AppState } from '.'

export interface RootState extends AppState {}

declare module '@nuxt/types' {
  interface Context {
    Store: Store<RootState>
  }
}
