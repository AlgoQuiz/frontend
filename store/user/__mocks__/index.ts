export const namespaced = true

export const state = () => ({})

export const mutations = {
  ON_AUTH_STATE_CHANGED_MUTATION: jest.fn(),
}

export const getters = {
  currentUser: jest.fn(),
  isLoggedIn: jest.fn(),
}

export const actions = {
  onAuthStateChangedAction: jest.fn(),
  signIn: jest.fn().mockReturnValue(Promise.resolve({})),
  signUpWithEmail: jest.fn().mockReturnValue(Promise.resolve({})),
  signOut: jest.fn().mockReturnValue(Promise.resolve({})),
}
