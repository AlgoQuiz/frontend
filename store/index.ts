import type { GetterTree, ActionTree, MutationTree } from 'vuex'
import { RootState } from '@/store/store.types'

export interface AppState {}

export const state = (): RootState => ({})

export const getters: GetterTree<RootState, RootState> = {}

export const mutations: MutationTree<RootState> = {}

export const actions: ActionTree<RootState, RootState> = {}

export const strict = false
