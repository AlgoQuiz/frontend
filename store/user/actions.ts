import firebase from 'firebase'
import { ActionTree } from 'vuex/types'

import { UserAuth, UserState } from '@/store/user/user.types'
import { RootState } from '../store.types'

interface UserDetails {
  email: string
  password: string
}

const actions: ActionTree<UserState, RootState> = {
  onAuthStateChangedAction: ({ commit }, { authUser, claims }: UserAuth) => {
    if (!authUser) {
      // Perform logout operations
    } else {
      // Do something with the authUser and the claims object...
    }
  },

  signIn: async (_, { email, password }: UserDetails) => {
    try {
      return await firebase.auth().signInWithEmailAndPassword(email, password)
    } catch (e) {
      console.error(e)
    }
  },

  signOut: async () => {
    try {
      await firebase.auth().signOut()
    } catch (e) {
      console.error(e)
    }
  },

  signUpWithEmail: async (_, { email, password }: UserDetails) => {
    try {
      return await firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
    } catch (e) {
      console.error(e)
    }
  },
}

export default actions
