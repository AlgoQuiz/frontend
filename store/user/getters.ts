import { GetterTree } from 'vuex/types'

import { RootState } from '../store.types'
import { UserState } from './user.types'

import { isEmptyObject } from '~/utils'

const getters: GetterTree<UserState, RootState> = {
  currentUser(state) {
    return state.data
  },

  isLoggedIn(_, getters) {
    return !isEmptyObject(getters.currentUser)
  },
}

export default getters
