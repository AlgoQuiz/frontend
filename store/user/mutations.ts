import { MutationTree } from 'vuex/types/index'
import { UserState, UserAuth } from './user.types'

const mutations: MutationTree<UserState> = {
  ON_AUTH_STATE_CHANGED_MUTATION: (
    state: UserState,
    { authUser }: UserAuth
  ) => {
    if (!authUser) {
      state.data = {}
      return
    }

    const { uid, email, emailVerified } = authUser

    state.data = { uid, email, emailVerified }
  },
}

export default mutations
