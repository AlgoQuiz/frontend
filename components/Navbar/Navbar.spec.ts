import { shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import * as user from '~/store/user'

import Navbar from './index.vue'

jest.mock('~/store/user')

describe('Components: Navbar', () => {
  let store: any

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        user,
      },
    })
  })

  const factory = (opts = {}) =>
    shallowMount(Navbar, {
      store,
      ...opts,
    })

  describe('snapshots', () => {
    it('matches default snapshot', () => {
      const wrapper = factory()

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
