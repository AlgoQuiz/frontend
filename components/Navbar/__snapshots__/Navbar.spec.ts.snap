// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Navbar snapshots matches default snapshot 1`] = `
"<bnavbar-stub variant=\\"primary\\" type=\\"dark\\" toggleable=\\"lg\\">
  <bnavbarbrand-stub to=\\"/\\">AlgoRace</bnavbarbrand-stub>
  <bnavbartoggle-stub target=\\"nav-collapse\\"></bnavbartoggle-stub>
  <bcollapse-stub isnav=\\"\\" id=\\"nav-collapse\\">
    <bnavbarnav-stub class=\\"ml-auto\\">
      <bnavitem-stub to=\\"/sign-in\\">Sign in</bnavitem-stub>
      <bnavitem-stub to=\\"/sign-up\\">Sign up</bnavitem-stub>
    </bnavbarnav-stub>
  </bcollapse-stub>
</bnavbar-stub>"
`;
